### PR TITLE
[gym] fix generating and exporting macOS `.app`

### LINKED ISSUE #15963: [gym] Export Developer ID macOS app fails (.app)
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
Note: I'm a little new to fastlane so please bare with me!

Using `gym` to build my macOS app, the command fails right after the export phase:

```
INFO [2020-02-02 18:34:26.67]: ▸ Exported MyApp to: /var/folders/sr/4s52t1q16yzgnn17nfwy_fv40000gn/T/gym_output20200202-85761-10rcbbx
INFO [2020-02-02 18:34:26.67]: ▸ ** EXPORT SUCCEEDED **
ERROR [2020-02-02 18:34:26.69]: The generated archive is invalid, this can have various reasons:
ERROR [2020-02-02 18:34:26.69]: Usually it's caused by the `Skip Install` option in Xcode, set it to `NO`
ERROR [2020-02-02 18:34:26.69]: For more information visit https://developer.apple.com/library/ios/technotes/tn2215/_index.html
ERROR [2020-02-02 18:34:26.69]: Also, make sure to have a valid code signing identity and provisioning profile installed
...
```

The export itself succeeds (a valid, correctly signed binary is produced) but fastlane appears unable to find the binary and assumes the export has failed.

This appears to be a bug in `gym/lib/gym/generators/package_command_generator_xcode7.rb` in `pkg_path`. The function assumes the export output has a suffix of `.pkg`. However, I am trying to export a Developer ID app, so the export produces a `.app`. 

If I 'hack' the `pkg_path` function and replace `.pkg` with `.app`, everything works perfectly.

##### Command executed
bundle exec fastlane myappbeta

##### Complete output when running fastlane, including the stack trace and command used
<details>
  <pre> [18:41:29]: Successfully exported and compressed dSYM file
[18:41:29]: Successfully exported the .app file:
[18:41:29]: ../MyApp Releases/MyApp/MyApp 4.1.10/MyApp 4.app
[18:41:29]: Generated plist file with the following values:
[18:41:29]: ▸ -----------------------------------------
[18:41:29]: ▸ {
[18:41:29]: ▸   "provisioningProfiles": {
[18:41:29]: ▸     "com.xxx.myapp": "match Direct com.xxx.myapp macos"
[18:41:29]: ▸   },
[18:41:29]: ▸   "method": "developer-id",
[18:41:29]: ▸   "signingStyle": "manual"
[18:41:29]: ▸ }
[18:41:29]: ▸ -----------------------------------------
[18:41:29]: $ /usr/bin/xcrun /Users/blah/Work/Xcode\ Projects/MyApp/vendor/bundle/ruby/2.6.0/gems/fastlane-2.141.0/gym/lib/assets/wrap_xcodebuild/xcbuild-safe.sh -exportArchive -exportOptionsPlist '/var/folders/sr/4s52t1q16yzgnn17nfwy_fv40000gn/T/gym_config20200202-95104-14e0ysk.plist' -archivePath /Users/blah/Library/Developer/Xcode/Archives/2020-02-02/MyApp\ 4\ 2020-02-02\ 18.41.09.xcarchive -exportPath '/var/folders/sr/4s52t1q16yzgnn17nfwy_fv40000gn/T/gym_output20200202-95104-1vwto0x'
[18:41:36]: The generated archive is invalid, this can have various reasons:
[18:41:36]: Usually it's caused by the `Skip Install` option in Xcode, set it to `NO`
[18:41:36]: For more information visit https://developer.apple.com/library/ios/technotes/tn2215/_index.html
[18:41:36]: Also, make sure to have a valid code signing identity and provisioning profile installed
[18:41:36]: Follow this guide to setup code signing https://docs.fastlane.tools/codesigning/GettingStarted/
[18:41:36]: If your intention was only to export an ipa be sure to provide a valid archive at the archive path.
[18:41:36]: This error might also happen if your workspace/project file is not in the root directory of your project.
[18:41:36]: To workaround that issue, you can wrap your calls to gym with
[18:41:36]: `Dir.chdir('../path/to/dir/containing/proj') do`
[18:41:36]: For an example you can check out
[18:41:36]: https://github.com/artsy/emission-nebula/commit/44fe51a7fea8f7d52f0f77d6c3084827fe5dd59e </pre>
</details>

### Environment

<details>
  <pre> <details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                                |
| --------------------------- | -------------------------------------------------------------------- |
| OS                          | 10.15.3                                                              |
| Ruby                        | 2.6.3                                                                |
| Bundler?                    | true                                                                 |
| Git                         | git version 2.21.1 (Apple Git-122.3)                                 |
| Installation Source         | ~/Work/Xcode Projects/MyApp/vendor/bundle/ruby/2.6.0/bin/fastlane |
| Host                        | Mac OS X 10.15.3 (19D76)                                             |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib       |
| OpenSSL Version             | LibreSSL 2.8.3                                                       |
| Is contained                | false                                                                |
| Is homebrew                 | false                                                                |
| Is installed via Fabric.app | false                                                                |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                          |
| Xcode Version               | 11.3.1                                                               |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# This file contains the fastlane.tools configuration
# You can find the documentation at https://docs.fastlane.tools
#
# For a list of all available actions, check out
#
#     https://docs.fastlane.tools/actions
#
# For a list of all available plugins, check out
#
#     https://docs.fastlane.tools/plugins/available-plugins
#

# Uncomment the line if you want fastlane to automatically update itself
# update_fastlane

# load variables from .env file in the root if it exists
if File.exist?('../.env')
  open('../.env', 'r').readlines.each do |l|
    kv = l.split('=')
    ENV[kv[0]] = kv[1].chomp
  end
end

default_platform(:mac)

platform :mac do
  desc "MyApp"

  # Configuration
  projectFilename = "MyApp.xcodeproj"

  # Before Integration Tests (Xcode Server Bot)
  lane :before_integration do
    sigh(platform: "macos",
         output_path: '/Library/Developer/XcodeServer/ProvisioningProfiles', 
         skip_install: true)
  end

  # Generate Beta and export it to Latest folder
  lane :myappbeta do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppPaddle"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/Paddle/" + versionedPath
    zipFilename = versionedFullPath + "/MyApp Paddle " + version + "-" + buildNumber + ".zip"
    appFilename = versionedFullPath + "/MyApp 4.app"

    # Set up certificates and provisioning profiles
    match(
      platform: "macos",
      type: "developer_id"
    )

    # Note: To make this work, currently need to edit /Users/blah/Work/Xcode Projects/MyApp/vendor/bundle/ruby/2.6.0/gems/fastlane-2.141.0/gym/lib/gym/generators/package_command_generator_xcode7.rb
    # around line 102 and replace all ".pkg" with ".app"
    build_mac_app(scheme: "MyApp Paddle Beta",
      output_directory: versionedFullPath,
      export_method: "developer-id",
      )

    notarize(
      package: appFilename,
      print_log: 'true',
      verbose: 'true'
    )      
      
    # Compress to a zip file
    zip(path: versionedFullPath + "/MyApp 4.app",
        output_path: zipFilename
    )
  end

  # Generate MAS Beta and export it to Latest folder
  lane :masbeta do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppMAS"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/MAS/" + versionedPath
    zipFilename = versionedFullPath + "/MyApp MAS " + version + "-" + buildNumber + ".zip"
    appFilename = versionedFullPath + "/MyApp 4.app"

    # Set up certificates and provisioning profiles
    match(
      platform: "macos",
      type: "developer_id"
    )
    build_mac_app(scheme: "MyApp MAS Beta",
        output_directory: versionedFullPath,
        export_method: "developer-id",
      )

    notarize(
      package: appFilename,
      print_log: 'true',
      verbose: 'true'
    )      
  
    # Compress to a zip file
    zip(path: versionedFullPath + "/MyApp 4.app",
        output_path: zipFilename
    )
  end  

  # Build the Paddle release version, export it and create the Disk Image distribution
  lane :buildpaddle do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppPaddle"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/Paddle/" + versionedPath
    appFilename = versionedFullPath + "/MyApp 4.app"
    zipFilename = versionedFullPath + "/MyApp " + version + ".zip"
    dmgFilename = versionedFullPath + "/MyApp " + version + ".dmg"
    dsymFilename = versionedFullPath + "/MyApp 4.app.dSYM.zip"

    # Set up certificates and provisioning profiles
    match(
      platform: "macos",
      type: "developer_id"
    )

    # Build
    gym(scheme: "MyApp Paddle",
        output_directory: versionedFullPath,
        export_method: "developer-id",
        # export_options: {
        #   method: "developer-id",
        #   provisioningProfiles: { 
        #     "com.xxx.myapp" => "Direct_com.xxx.myapp.provisionprofile"
        #   },
        # },
        # export_xcargs: "-allowProvisioningUpdates"
      )
    
    # Unfortunately, Fastlane doesn't properly export the built product (it simply copies it from the xcarchive) and it doesn't
    # use the provided provisioning profile for signing.
    # UI.important "\n\nUnfortunately, right now, this is as far as Fastlane can go.\nNext step is to visit the Organiser, Notarise the app and export the notarised copy to " +appFilename + ".\nOnce that is done, run the paddlerelease lane to continue with distribution."
    # Notarize the app
    notarize(
      package: appFilename,
      print_log: 'true',
      verbose: 'true'
    )      
  end

  # Peform the release of the Paddle version. Notarize, ZIP, upload to AppCenter, DropDMG and upload to xxx.
  lane :releasepaddle do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppPaddle"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/Paddle/" + versionedPath
    appFilename = versionedFullPath + "/MyApp 4.app"
    zipFilename = versionedFullPath + "/MyApp " + version + ".zip"
    dmgFilename = versionedFullPath + "/MyApp " + version + ".dmg"
    dsymFilename = versionedFullPath + "/MyApp 4.app.dSYM.zip"
    
    # Compress to a zip file
    begin
      # Put this item in an error handler because it may fail (and otherwise cause execution to stop)
      sh("rm", "../" + zipFilename)
    rescue => ex
      UI.error(ex)
    end    
    zip(path: appFilename,
        output_path: zipFilename
    )

    # Upload to MS AppCenter
    appcenter_upload(
      api_token: ENV["APPCENTER_API_TOKEN"],
      owner_name: ENV["APPCENTER_OWNER_NAME"],
      app_name: ENV["APPCENTER_MACOS_APP_NAME_PADDLE"],
      app_os: 'MacOS',  
      file: appFilename,
      dsym: dsymFilename,
    )    

    # Generate custom Disk Image
    begin
      # Put this item in an error handler because it may fail (and otherwise cause execution to stop)
      sh("rm", "../" + dmgFilename)
    rescue => ex
      UI.error(ex)
    end    
    sh("dropdmg", 
       "--license-folder", 
       "./DropDMG/Default", 
       "--layout-folder", 
       "DropDMG", 
       "--config-name", 
       "MyApp 4", 
       "--base-name", 
       versionedPath, 
       "../" + versionedFullPath + "/MyApp 4.app"
    )

    # Upload disk image to xxx.xxx.com
    scp(host: "xxx.xxx.com",
        username: "username",
        upload: {
          src: dmgFilename,
          dst: "/var/www/xxx/releases"
        }
    )
    # Upload ZIP to xxx.xxx.com
    scp(host: "xxx.xxx.com",
        username: "username",
        upload: {
          src: zipFilename,
          dst: "/var/www/xxx/releases"
        }
    )
  end

  # Generate MAS Beta and export it to Latest folder
  lane :masrelease do
    buildNumber = get_info_plist_value(path: "./Info.plist", key: "CFBundleVersion")
    version = get_version_number(xcodeproj: projectFilename,
                                  target: "MyAppMAS"
    )
    versionedPath = "MyApp " + version
    versionedFullPath = "../MyApp Releases/MAS/" + versionedPath
    zipFilename = versionedFullPath + "/MyApp MAS " + version + "-" + buildNumber + ".zip"
    dsymFilename = versionedFullPath + "/MyApp 4.app.dSYM.zip"
    appFilename = versionedFullPath + "/MyApp 4.app"

    # Set up certificates and provisioning profiles
    match(
      platform: "macos",
      type: "appstore"
    )

    build_mac_app(scheme: "MyApp MAS",
      output_directory: versionedFullPath,
      # export_options: {
      #   method: "package",
      #   provisioningProfiles: { 
      #     "com.xxx.myapp" => "AppStore_com.xxx.myapp.provisionprofile"
      #   }
      # },
      # export_xcargs: "-allowProvisioningUpdates"
    )

    ### Note, Gym does not yet support exporting a .pkg required for uploading to the AppStore.
    ### Once Gym is complete, go to the Xcode Organizer and submit the build to the App Store.
    ### Keep an eye on this PR that adds support for Mac Pkg and AppStore uploads:
    ### https://github.com/fastlane/fastlane/pull/12195
    deliver(
      platform: "osx",
      skip_screenshots: "true",
      skip_metadata: "true",
      run_precheck_before_submit: "false",
      force: "true"
    )      

    # Upload to MS AppCenter
    appcenter_upload(
      api_token: ENV["APPCENTER_API_TOKEN"],
      owner_name: ENV["APPCENTER_OWNER_NAME"],
      app_name: ENV["APPCENTER_MACOS_APP_NAME_MAS"],
      app_os: 'MacOS',  
      file: appFilename,
      dsym: dsymFilename,
    )    
  end  

end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
app_identifier("com.xxx.myapp") # The bundle identifier of your app
apple_id("emailaddress") # Your Apple email address

# For more information about the Appfile, see:
#     https://docs.fastlane.tools/advanced/#appfile

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.141.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                    | Version | Update-Status |
| ------------------------- | ------- | ------------- |
| fastlane-plugin-appcenter | 1.7.1   | ✅ Up-To-Date  |
| fastlane-plugin-notarize  | 1.1.2   | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.3.0        |
| bundler                   | 1.17.2       |
| etc                       | 1.0.1        |
| forwardable               | 1.2.0        |
| CFPropertyList            | 3.0.2        |
| public_suffix             | 2.0.5        |
| addressable               | 2.7.0        |
| atomos                    | 0.1.3        |
| babosa                    | 1.0.3        |
| claide                    | 1.0.3        |
| colored                   | 1.2          |
| colored2                  | 3.1.2        |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| digest-crc                | 0.4.1        |
| unf_ext                   | 0.0.7.6      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| dotenv                    | 2.7.5        |
| emoji_regex               | 1.0.1        |
| excon                     | 0.72.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.17.3       |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.13.1       |
| fastimage                 | 2.1.7        |
| gh_inspector              | 1.1.3        |
| jwt                       | 2.1.0        |
| memoist                   | 0.16.2       |
| multi_json                | 1.14.1       |
| os                        | 1.0.1        |
| signet                    | 0.12.0       |
| googleauth                | 0.10.0       |
| httpclient                | 2.8.3        |
| mini_mime                 | 1.0.2        |
| uber                      | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| google-api-client         | 0.36.4       |
| google-cloud-env          | 1.3.0        |
| google-cloud-errors       | 1.0.0        |
| google-cloud-core         | 1.5.0        |
| google-cloud-storage      | 1.25.1       |
| json                      | 2.3.0        |
| mini_magick               | 4.10.1       |
| multi_xml                 | 0.6.0        |
| plist                     | 3.5.0        |
| rubyzip                   | 1.3.0        |
| security                  | 0.1.3        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.7        |
| slack-notifier            | 2.3.2        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.6.1        |
| terminal-table            | 1.8.0        |
| tty-screen                | 0.7.0        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| word_wrap                 | 1.0.0        |
| nanaimo                   | 0.2.6        |
| xcodeproj                 | 1.14.0       |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| xcpretty-travis-formatter | 1.0.0        |
| fastlane-plugin-appcenter | 1.7.1        |
| fastlane-plugin-notarize  | 1.1.2        |
| net-ssh                   | 5.2.0        |
| net-scp                   | 2.0.0        |
</details>

*generated on:* **2020-02-02**
</details> </pre>
</details>

### LINKED ISSUE #17706: Setting 'Skip Install' to NO generates a invalid archive MacOS
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

When running gym() on a MacOS X project. I am getting an error event though it success full creates .app:

[13:38:04]: The generated archive is invalid, this can have various reasons:
[13:38:04]: Usually it's caused by the `Skip Install` option in Xcode, set it to `NO`
[13:38:04]: For more information visit https://developer.apple.com/library/ios/technotes/tn2215/_index.html
[13:38:04]: Also, make sure to have a valid code signing identity and provisioning profile installed
[13:38:04]: Follow this guide to setup code signing https://docs.fastlane.tools/codesigning/GettingStarted/
[13:38:04]: If your intention was only to export an ipa be sure to provide a valid archive at the archive path.
[13:38:04]: This error might also happen if your workspace/project file is not in the root directory of your project.
[13:38:04]: To workaround that issue, you can wrap your calls to gym with
[13:38:04]: `Dir.chdir('../path/to/dir/containing/proj') do`
[13:38:04]: For an example you can check out
[13:38:04]: https://github.com/artsy/emission-nebula/commit/44fe51a7fea8f7d52f0f77d6c3084827fe5dd59e

If I set the Skip Install to NO then the archive does not create Products/Application/... I have already checked to make sure the Install Directory is ${LOCAL_APP_DIR}. 

##### Command executed

fastlane

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre>
[13:38:04]: The generated archive is invalid, this can have various reasons:
[13:38:04]: Usually it's caused by the `Skip Install` option in Xcode, set it to `NO`
[13:38:04]: For more information visit https://developer.apple.com/library/ios/technotes/tn2215/_index.html
[13:38:04]: Also, make sure to have a valid code signing identity and provisioning profile installed
[13:38:04]: Follow this guide to setup code signing https://docs.fastlane.tools/codesigning/GettingStarted/
[13:38:04]: If your intention was only to export an ipa be sure to provide a valid archive at the archive path.
[13:38:04]: This error might also happen if your workspace/project file is not in the root directory of your project.
[13:38:04]: To workaround that issue, you can wrap your calls to gym with
[13:38:04]: `Dir.chdir('../path/to/dir/containing/proj') do`
[13:38:04]: For an example you can check out
[13:38:04]: https://github.com/artsy/emission-nebula/commit/44fe51a7fea8f7d52f0f77d6c3084827fe5dd59e
  </pre>
</details>

### Environment

<details>
  <pre> 
<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 11.0.1                                                         |
| Ruby                        | 2.6.3                                                          |
| Bundler?                    | false                                                          |
| Git                         | git version 2.29.2                                             |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | macOS 11.0.1 (20B29)                                           |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 12.2                                                           |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# This file contains the fastlane.tools configuration
# You can find the documentation at https://docs.fastlane.tools
#
# For a list of all available actions, check out
#
#     https://docs.fastlane.tools/actions
#
# For a list of all available plugins, check out
#
#     https://docs.fastlane.tools/plugins/available-plugins
#

# Uncomment the line if you want fastlane to automatically update itself
# update_fastlane

default_platform(:mac)

platform :mac do
  desc "Description of what the lane does"
  lane :custom_lane do
    gym(
      scheme: "4D Launcher",
      configuration: "Release",
      clean: true, 
      output_directory: "./build",
      export_method: "development")
  end
end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
# app_identifier("[[APP_IDENTIFIER]]") # The bundle identifier of your app
# apple_id("[[APPLE_ID]]") # Your Apple email address


# For more information about the Appfile, see:
#     https://docs.fastlane.tools/advanced/#appfile

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.168.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.3.0        |
| slack-notifier            | 2.3.2        |
| atomos                    | 0.1.3        |
| claide                    | 1.0.3        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.3.0        |
| xcodeproj                 | 1.18.0       |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.7.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| public_suffix             | 4.0.6        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.8.1        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| babosa                    | 1.0.3        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| excon                     | 0.76.0       |
| faraday                   | 1.0.1        |
| unf_ext                   | 0.0.7.7      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.7        |
| faraday_middleware        | 1.0.0        |
| fastimage                 | 2.2.0        |
| gh_inspector              | 1.1.3        |
| mini_magick               | 4.10.1       |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| dotenv                    | 2.7.6        |
| bundler                   | 1.17.2       |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| jwt                       | 2.2.2        |
| uber                      | 0.1.0        |
| declarative               | 0.0.20       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| multi_json                | 1.15.0       |
| signet                    | 0.14.0       |
| memoist                   | 0.16.2       |
| os                        | 1.1.1        |
| googleauth                | 0.13.1       |
| httpclient                | 2.8.3        |
| google-api-client         | 0.38.0       |
| google-cloud-env          | 1.3.3        |
| google-cloud-errors       | 1.0.1        |
| google-cloud-core         | 1.5.0        |
| rake                      | 13.0.1       |
| digest-crc                | 0.6.1        |
| google-cloud-storage      | 1.28.0       |
| emoji_regex               | 3.0.0        |
| jmespath                  | 1.4.0        |
| aws-eventstream           | 1.1.0        |
| aws-sigv4                 | 1.2.2        |
| aws-sdk-core              | 3.105.0      |
| aws-sdk-kms               | 1.37.0       |
| aws-sdk-s3                | 1.79.1       |
| CFPropertyList            | 3.0.2        |
| json                      | 2.3.1        |
| aws-partitions            | 1.365.0      |
| forwardable               | 1.2.0        |
| logger                    | 1.3.0        |
| stringio                  | 0.0.2        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.1.0        |
| strscan                   | 1.0.0        |
| date                      | 2.0.0        |
| fileutils                 | 1.1.0        |
| etc                       | 1.0.1        |
| io-console                | 0.4.7        |
| zlib                      | 1.0.0        |
| libxml-ruby               | 3.1.0        |
| rexml                     | 3.1.9        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.4.2        |
</details>

*generated on:* **2020-12-01**
</details>
</pre>
</details>

### PR DESCRIPTION
**Pull request by EpaL**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Previously, gym would only correct detect the generated build if it was a .pkg file. If the build output a macOS .app file, gym would fail because it could not find the built product. The workaround was to set skip_package_pkg: true, but this would skip the export step, resulting in an incorrectly signed output.

Resolves #15963
Resolves #17706

### Description
Added support for detecting when the output of a macOS build is a .app (instead of .pkg). Previously, if output was a .app and export options were set, Fastlane would fail because it was expected a .pkg output. 

Tested and confirmed working by myself and others in https://github.com/fastlane/fastlane/issues/15963#issuecomment-583363525. That being said, the code is very rudimentary and could definitely use a more expert pair of eyes on it.

### Testing Steps

To test this branch, modify your Gemfile as:

```ruby
gem "fastlane", :git => "https://github.com/EpaL/fastlane.git", :branch => "master"
```

And run `bundle install` to apply the changes.

### CODE REVIEW

**Comment by EpaL** [comment]
@googlebot I signed it!

File: gym/lib/gym/generators/package_command_generator_xcode7.rb [hunk 1]
```diff
@@ -98,6 +99,10 @@ def pkg_path
           path = files.min_by(&:length)
           Gym.cache[:pkg_path] = File.join(temporary_output_path, "#{Gym.config[:output_name]}.pkg")
           FileUtils.cp(path, Gym.cache[:pkg_path]) unless File.expand_path(path).casecmp(File.expand_path(Gym.cache[:pkg_path]).downcase).zero?
+        elsif appPath
+          # Try to find .app file in the output directory. This is used when macOS is set and .app is being generated.
+          Gym.cache[:pkg_path] = File.join(temporary_output_path, "#{Gym.config[:output_name]}.app")
```
Comment (rogerluan) [inline gym/lib/gym/generators/package_command_generator_xcode7.rb:104]: Minor semantic comment: here we're attributing a `.app` to something called `pkg_path`. Looks like re-using the same cache key as the pkg itself simplify things a bit here (and avoid other issues probably), so I'd be inclined to name it to be more generic (such as output_path, or binary_path, or something else that doesn't necessarily specify the file extension in its name).

**Review by rogerluan — CHANGES_REQUESTED** [review]
Hey @EpaL 👋 

Thanks for your first contribution! 🚀 
While I haven't personally tested your PR, I saw some testimonials in the referred issue of people confirming it works. Moreover, your fix is clean enough so that it probably wouldn't ever cause any regressions 🙌 so I think it's safe.

Ah, I did some grooming to your PR description (linked to the issue and added testing steps) 🤗 

Could you update your branch with the latest master? 🙏 Looks like it's 300+ commits behind (which's probably the reason why CI is failing)

File: gym/lib/gym/generators/package_command_generator_xcode7.rb (hunk 1 above)
Comment (EpaL) [reply gym/lib/gym/generators/package_command_generator_xcode7.rb]: Thanks, done and done! I tested the last changes on my build and everything seems to work.

Comment (rogerluan) [reply gym/lib/gym/generators/package_command_generator_xcode7.rb]: Thanks!

**Review by rogerluan — APPROVED** [review]
🎉 Thanks for addressing the changes I requested!

**Comment by appfrilans** [comment]
Any ETA on this being published in a public release?

**Comment by DavidBrunow** [comment]
I was looking through the open pull requests and noticed that this one is similar to my pull request here: https://github.com/fastlane/fastlane/pull/18016/files

Don't want to give you additional work but if you have some spare time would you mind checking if the fix in my pull request resolves your issue @EpaL?

**Comment by wuftymerguftyguff** [comment]
FWIW the testing branch from this PR fixes this form me too.

**Comment by appfrilans** [comment]
Whop! 🙌 Everything approved? Could this perhaps be in the next release??

**Comment by appfrilans** [comment]
Does joshdholtz need to approve as well?

**Comment by rogerluan** [comment]
Yup ☝️ 🙇  @joshdholtz friendly ping 🤗 this has been requested by many! Thanks in advance!

**Comment by jason-gabriele** [comment]
Would love to see this in the next release. Thanks!

**Review by joshdholtz — APPROVED** [review]
This works so great! I have a macOS app that I release on App Store and outside the App Store and I had to use that `skip_package_pkg` to get the outside App Store to work but now I don’t have too 🙃 

Thank you so much for looking into this and fixing it! I really appreciate this ❤️

**Comment by EpaL** [comment]
@joshdholtz, that's awesome! I'm in the same boat so happy to see this going out. Thank you thank you thank you to all who helped make it happen! 🎉

**Comment by alinradut** [comment]
Just tested this using the 2.179.0 release, it works as expected!

**Pull request merged**